File: internal/session/prepared.go
package session

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"

	"golang.org/x/sys/unix"
)

// Prepared pins one verified session file descriptor and its runtime metadata.
// Preparing is read-only. Activate consumes the handle and transfers that exact
// descriptor to a Store, which may then perform the normal documented repairs.
type Prepared struct {
	mu             sync.Mutex
	path           string
	file           *os.File
	fileInfo       os.FileInfo
	info           SessionInfo
	verifyIdentity func(os.FileInfo) error
}

func Prepare(ctx context.Context, path string) (*Prepared, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	file, fileInfo, err := openPreparedSessionFileNoFollow(path)
	if err != nil {
		return nil, err
	}
	return prepareOpenedSession(ctx, path, file, fileInfo, func(prepared os.FileInfo) error {
		return verifyPreparedPathIdentity(path, prepared)
	})
}

// PrepareListed opens only a candidate returned for the expected session root
// and workspace. Every attacker-controlled component below root is traversed
// with openat and O_NOFOLLOW before the exact regular-file descriptor is pinned.
func PrepareListed(ctx context.Context, root, workspace, path string) (*Prepared, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	rootPath, workspaceCanonical, basename, candidatePath, err := validateListedCandidatePath(root, workspace, path)
	if err != nil {
		return nil, err
	}
	file, fileInfo, err := openListedPreparedSessionFile(rootPath, workspaceCanonical, basename)
	if err != nil {
		return nil, err
	}
	prepared, err := prepareOpenedSession(ctx, candidatePath, file, fileInfo, func(expected os.FileInfo) error {
		current, currentInfo, openErr := openListedPreparedSessionFile(rootPath, workspaceCanonical, basename)
		if openErr != nil {
			return fmt.Errorf("%w: prepared listed session path identity changed before activation", ErrInvalidSession)
		}
		defer current.Close()
		if !os.SameFile(expected, currentInfo) {
			return fmt.Errorf("%w: prepared listed session path identity changed before activation", ErrInvalidSession)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	candidateWorkspace, err := canonicalWorkspace(prepared.info.CWD)
	if err != nil || candidateWorkspace != workspaceCanonical {
		_ = prepared.Close()
		return nil, fmt.Errorf("%w: listed session workspace does not match expected workspace", ErrInvalidSession)
	}
	return prepared, nil
}

func prepareOpenedSession(ctx context.Context, path string, file *os.File, fileInfo os.FileInfo, verifyIdentity func(os.FileInfo) error) (*Prepared, error) {
	closeOnError := func(prepareErr error) (*Prepared, error) {
		if closeErr := file.Close(); closeErr != nil {
			prepareErr = errors.Join(prepareErr, fmt.Errorf("close prepared session file: %w", closeErr))
		}
		return nil, prepareErr
	}

	info, _, err := inspectOpenedSession(ctx, path, file, fileInfo)
	if err != nil {
		return closeOnError(err)
	}
	if err := ctx.Err(); err != nil {
		return closeOnError(err)
	}
	return &Prepared{path: path, file: file, fileInfo: fileInfo, info: info, verifyIdentity: verifyIdentity}, nil
}

func (p *Prepared) Info() SessionInfo {
	if p == nil {
		return SessionInfo{}
	}
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.info
}

// Activate consumes the prepared handle. On success the returned Store owns
// the descriptor; on failure the descriptor is closed.
func (p *Prepared) Activate(ctx context.Context) (*Store, []Warning, error) {
	if p == nil {
		return nil, nil, errors.New("prepared session is no longer available")
	}
	p.mu.Lock()
	file := p.file
	p.file = nil
	path := p.path
	fileInfo := p.fileInfo
	preparedInfo := p.info
	verifyIdentity := p.verifyIdentity
	p.mu.Unlock()
	if file == nil {
		return nil, nil, errors.New("prepared session is no longer available")
	}
	closeOnError := func(activateErr error) (*Store, []Warning, error) {
		if closeErr := file.Close(); closeErr != nil {
			activateErr = errors.Join(activateErr, fmt.Errorf("close prepared session file: %w", closeErr))
		}
		return nil, nil, activateErr
	}
	if err := ctx.Err(); err != nil {
		return closeOnError(err)
	}
	if verifyIdentity != nil {
		if err := verifyIdentity(fileInfo); err != nil {
			return closeOnError(err)
		}
	}
	currentFileInfo, err := file.Stat()
	if err != nil {
		return closeOnError(fmt.Errorf("stat prepared session file: %w", err))
	}
	if !os.SameFile(fileInfo, currentFileInfo) {
		return closeOnError(fmt.Errorf("%w: prepared session file identity changed", ErrInvalidSession))
	}
	currentInfo, _, err := inspectOpenedSession(ctx, path, file, currentFileInfo)
	if err != nil {
		return closeOnError(err)
	}
	if !samePreparedMetadata(preparedInfo, currentInfo) {
		return closeOnError(fmt.Errorf("%w: prepared session metadata changed before activation", ErrInvalidSession))
	}

	// openStoreFromFile consumes file on both success and failure.
	return openStoreFromFile(file, path)
}

// Close abandons a prepared handle. It is safe to call after activation and is
// idempotent.
func (p *Prepared) Close() error {
	if p == nil {
		return nil
	}
	p.mu.Lock()
	file := p.file
	p.file = nil
	p.mu.Unlock()
	if file == nil {
		return nil
	}
	if err := file.Close(); err != nil {
		return fmt.Errorf("close prepared session file: %w", err)
	}
	return nil
}

func validateListedCandidatePath(root, workspace, path string) (string, string, string, string, error) {
	rootPath, err := filepath.Abs(root)
	if err != nil {
		return "", "", "", "", fmt.Errorf("resolve session root path: %w", err)
	}
	rootPath = filepath.Clean(rootPath)
	workspaceCanonical, err := canonicalWorkspace(workspace)
	if err != nil {
		return "", "", "", "", err
	}
	key, err := workspaceKey(workspaceCanonical)
	if err != nil {
		return "", "", "", "", err
	}
	candidatePath, err := filepath.Abs(path)
	if err != nil {
		return "", "", "", "", fmt.Errorf("resolve listed session path: %w", err)
	}
	candidatePath = filepath.Clean(candidatePath)
	basename := filepath.Base(candidatePath)
	expected := filepath.Join(rootPath, key, basename)
	if candidatePath != expected || basename == "." || basename == ".." || filepath.Ext(basename) != ".jsonl" {
		return "", "", "", "", fmt.Errorf("%w: listed session candidate is outside the expected workspace directory", ErrInvalidSession)
	}
	return rootPath, workspaceCanonical, basename, candidatePath, nil
}

func openListedPreparedSessionFile(root, workspace, basename string) (*os.File, os.FileInfo, error) {
	rootDir, err := openSessionRootNoFollow(root)
	if err != nil {
		return nil, nil, err
	}
	defer rootDir.Close()
	workspaceDir, _, exists, err := openWorkspaceSessionDirectoryNoFollow(rootDir, root, workspace)
	if err != nil {
		return nil, nil, err
	}
	if !exists {
		return nil, nil, fmt.Errorf("%w: listed session directory no longer exists", ErrInvalidSession)
	}
	defer workspaceDir.Close()

	file, err := openPathAtNoFollow(workspaceDir, basename, unix.O_RDWR, 0)
	if err != nil {
		if err == unix.ELOOP {
			return nil, nil, fmt.Errorf("%w: listed session file is a symlink", ErrInvalidSession)
		}
		return nil, nil, fmt.Errorf("open listed session file: %w", err)
	}
	info, err := file.Stat()
	if err != nil {
		_ = file.Close()
		return nil, nil, fmt.Errorf("stat listed session file: %w", err)
	}
	if !info.Mode().IsRegular() {
		_ = file.Close()
		return nil, nil, fmt.Errorf("%w: listed session file is not a regular file", ErrInvalidSession)
	}
	return file, info, nil
}

func openPreparedSessionFileNoFollow(path string) (*os.File, os.FileInfo, error) {
	file, err := openPathNoFollow(path, unix.O_RDWR, 0)
	if err != nil {
		if err == unix.ELOOP {
			return nil, nil, fmt.Errorf("%w: session file is a symlink", ErrInvalidSession)
		}
		return nil, nil, fmt.Errorf("open session file: %w", err)
	}
	info, err := file.Stat()
	if err != nil {
		_ = file.Close()
		return nil, nil, fmt.Errorf("stat session file: %w", err)
	}
	if !info.Mode().IsRegular() {
		_ = file.Close()
		return nil, nil, fmt.Errorf("%w: session file is not a regular file", ErrInvalidSession)
	}
	return file, info, nil
}

func verifyPreparedPathIdentity(path string, prepared os.FileInfo) error {
	current, err := os.Lstat(path)
	if err != nil || !current.Mode().IsRegular() || !os.SameFile(prepared, current) {
		return fmt.Errorf("%w: prepared session path identity changed before activation", ErrInvalidSession)
	}
	return nil
}

func samePreparedMetadata(prepared, current SessionInfo) bool {
	return prepared.ID == current.ID &&
		prepared.CWD == current.CWD &&
		prepared.Profile == current.Profile &&
		prepared.Provider == current.Provider &&
		prepared.Model == current.Model &&
		prepared.Created.Equal(current.Created)
}

File: crates/otto/src/session/prepared.rs
//! Preparing a session for activation, and archiving one.
//!
//! Port of `internal/session/prepared.go` and `internal/session/archive.go`.
//!
//! Ownership: a [`Prepared`] owns one verified descriptor until
//! [`Prepared::activate`] transfers it to a [`Store`] or [`Prepared::close`]
//! releases it; both are idempotent and neither leaks the descriptor on
//! failure. Concurrency: the handle is behind a mutex, so racing `activate`
//! calls leave exactly one winner. Errors: an identity that changed between
//! preparation and activation is rejected rather than used.

use std::fs::{File, Metadata};
use std::path::Path;
use std::sync::Mutex;

use otto_core::session::{PiError, SessionInfo, Warning};

use super::fsops;
use super::list::{
    inspect_opened_session, open_session_file_read_only_no_follow_at, open_session_root_no_follow,
    open_workspace_session_directory_no_follow,
};
use super::store::Store;

const ARCHIVE_DIRECTORY_NAME: &str = "archive";

/// How a prepared handle re-checks, at activation time, that the path it was
/// opened through still names the same file.
#[derive(Debug)]
enum Identity {
    /// Re-`lstat` the path the caller gave.
    Path(String),
    /// Re-open the candidate through the session root and compare.
    Listed {
        root: String,
        workspace: String,
        basename: String,
    },
}

/// One verified session file descriptor and the metadata read from it.
#[derive(Debug)]
pub struct Prepared {
    path: String,
    handle: Mutex<Option<(File, Metadata)>>,
    info: SessionInfo,
    identity: Identity,
}

/// A completed archive move.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ArchiveResult {
    pub path: String,
    pub id: String,
}

impl Prepared {
    /// Pins the session file at `path`, read-write, refusing a symlink.
    pub fn prepare(path: &Path) -> Result<Self, PiError> {
        let (file, metadata) = open_prepared_session_file_no_follow(path)?;
        let path = path.to_string_lossy().into_owned();
        Self::from_opened(path.clone(), file, metadata, Identity::Path(path))
    }

    /// Pins a candidate that `list` returned for this root and workspace.
    ///
    /// Every attacker-controlled component below the root is traversed with
    /// `openat` and `O_NOFOLLOW` before the descriptor is pinned, and the
    /// session's recorded workspace must match the expected one.
    pub fn prepare_listed(root: &Path, workspace: &str, path: &Path) -> Result<Self, PiError> {
        let (root_path, workspace_canonical, basename, candidate_path) =
            validate_listed_candidate_path(root, workspace, path)?;
        let (file, metadata) =
            open_listed_prepared_session_file(&root_path, &workspace_canonical, &basename)?;
        let prepared = Self::from_opened(
            candidate_path,
            file,
            metadata,
            Identity::Listed {
                root: root_path,
                workspace: workspace_canonical.clone(),
                basename,
            },
        )?;
        match fsops::canonical_workspace(Path::new(&prepared.info.cwd)) {
            Ok(candidate) if candidate == workspace_canonical => Ok(prepared),
            _ => {
                prepared.close()?;
                Err(PiError::invalid(
                    "listed session workspace does not match expected workspace",
                ))
            }
        }
    }

    fn from_opened(
        path: String,
        mut file: File,
        metadata: Metadata,
        identity: Identity,
    ) -> Result<Self, PiError> {
        let (info, _) = inspect_opened_session(&path, &mut file, &metadata)?;
        Ok(Self {
            path,
            handle: Mutex::new(Some((file, metadata))),
            info,
            identity,
        })
    }

    /// The listing row read while preparing.
    pub fn info(&self) -> SessionInfo {
        self.info.clone()
    }

    /// Consumes the handle and returns a store owning the same descriptor.
    ///
    /// Fails, releasing the descriptor, when the path now names a different
    /// file or the session's identifying metadata changed since preparation.
    pub fn activate(self) -> Result<(Store, Vec<Warning>), PiError> {
        let taken = self
            .handle
            .lock()
            .map_err(|_| PiError::other("prepared session mutex is poisoned"))?
            .take();
        let Some((mut file, metadata)) = taken else {
            return Err(PiError::other("prepared session is no longer available"));
        };
        self.verify_identity(&metadata)?;

        let current_metadata = file
            .metadata()
            .map_err(|error| PiError::other(format!("stat prepared session file: {error}")))?;
        if fsops::file_identity(&metadata) != fsops::file_identity(&current_metadata) {
            return Err(PiError::invalid("prepared session file identity changed"));
        }
        let (current_info, _) = inspect_opened_session(&self.path, &mut file, &current_metadata)?;
        if !same_prepared_metadata(&self.info, &current_info) {
            return Err(PiError::invalid(
                "prepared session metadata changed before activation",
            ));
        }
        Store::from_file(file, &self.path)
    }

    fn verify_identity(&self, prepared: &Metadata) -> Result<(), PiError> {
        const CHANGED: &str = "prepared session path identity changed before activation";
        match &self.identity {
            Identity::Path(path) => {
                let current = std::fs::symlink_metadata(path);
                match current {
                    Ok(current)
                        if current.is_file()
                            && fsops::file_identity(prepared) == fsops::file_identity(&current) =>
                    {
                        Ok(())
                    }
                    _ => Err(PiError::invalid(CHANGED)),
                }
            }
            Identity::Listed {
                root,
                workspace,
                basename,
            } => {
                let Ok((_, current)) = open_listed_prepared_session_file(root, workspace, basename)
                else {
                    return Err(PiError::invalid(
                        "prepared listed session path identity changed before activation",
                    ));
                };
                if fsops::file_identity(prepared) != fsops::file_identity(&current) {
                    return Err(PiError::invalid(
                        "prepared listed session path identity changed before activation",
                    ));
                }
                Ok(())
            }
        }
    }

    /// Abandons the handle. Safe after activation and idempotent.
    pub fn close(&self) -> Result<(), PiError> {
        self.handle
            .lock()
            .map_err(|_| PiError::other("prepared session mutex is poisoned"))?
            .take();
        Ok(())
    }
}

/// Moves one active session file into the workspace's `archive/` directory.
///
/// The source must be a direct child of the workspace-key directory and a
/// valid Pi v3 session whose recorded workspace matches the current canonical
/// workspace. The move is atomic and never replaces an existing archived
/// session; on success the file exists only under `archive/`.
pub fn archive(root: &Path, workspace: &str, path: &Path) -> Result<ArchiveResult, PiError> {
    let (_, workspace_canonical, basename, _) =
        validate_listed_candidate_path(root, workspace, path)?;

    let root_dir = open_session_root_no_follow(root)?;
    let Some((workspace_dir, directory)) =
        open_workspace_session_directory_no_follow(&root_dir, root, &workspace_canonical)?
    else {
        return Err(PiError::invalid(
            "listed session directory no longer exists",
        ));
    };

    let (mut file, metadata, candidate_path) =
        open_session_file_read_only_no_follow_at(&workspace_dir, &directory, &basename)?;
    let (session_info, _) = inspect_opened_session(&candidate_path, &mut file, &metadata)?;
    match fsops::canonical_workspace(Path::new(&session_info.cwd)) {
        Ok(recorded) if recorded == workspace_canonical => {}
        _ => {
            return Err(PiError::invalid(
                "session workspace does not match expected workspace",
            ));
        }
    }

    match std::fs::symlink_metadata(&candidate_path) {
        Ok(current)
            if current.is_file()
                && fsops::file_identity(&metadata) == fsops::file_identity(&current) => {}
        _ => {
            return Err(PiError::invalid(
                "session path identity changed before archive",
            ));
        }
    }

    let archive_dir = ensure_archive_directory(&workspace_dir)?;
    let destination = Path::new(&directory)
        .join(ARCHIVE_DIRECTORY_NAME)
        .join(&basename);
    ensure_archive_destination_absent(&archive_dir, &basename)?;

    // RENAME_EXCL makes the move atomic and refuses to replace an existing
    // destination, closing the check-then-rename race.
    fsops::rename_excl(Path::new(&candidate_path), &destination)
        .map_err(|error| PiError::other(format!("archive session file: {error}")))?;
    // Archiving ends the session, so its outstanding timers end with it: the
    // sidecar is removed rather than moved. Best-effort, because the session
    // file has already moved and there is no state left to roll back to. A
    // live session also clears its in-process timers in
    // `app::Controller::archive_current_session`.
    let _ = std::fs::remove_file(Path::new(&candidate_path).with_extension("reminders.json"));
    Ok(ArchiveResult {
        path: destination.to_string_lossy().into_owned(),
        id: session_info.id,
    })
}

/// Opens the verified `archive/` directory, creating it 0700 when missing.
fn ensure_archive_directory(workspace_dir: &fsops::Dir) -> Result<fsops::Dir, PiError> {
    if let Err(error) = fsops::mkdir_at(workspace_dir, ARCHIVE_DIRECTORY_NAME, 0o700)
        && error.raw_os_error() != Some(libc::EEXIST)
    {
        return Err(PiError::other(format!("create archive directory: {error}")));
    }
    let archive_dir =
        fsops::open_dir_at_no_follow(workspace_dir, ARCHIVE_DIRECTORY_NAME).map_err(|error| {
            if fsops::is_eloop(&error) {
                PiError::invalid("archive directory is a symlink")
            } else if fsops::is_enotdir(&error) {
                PiError::invalid("archive path is not a directory")
            } else {
                PiError::other(format!("open archive directory: {error}"))
            }
        })?;
    fsops::fchmod_dir(&archive_dir)
        .map_err(|error| PiError::other(format!("chmod archive directory: {error}")))?;
    Ok(archive_dir)
}

/// Rejects an archive move whose destination already exists.
fn ensure_archive_destination_absent(
    archive_dir: &fsops::Dir,
    basename: &str,
) -> Result<(), PiError> {
    match fsops::exists_at_no_follow(archive_dir, basename) {
        Ok(true) => Err(PiError::invalid("archive destination already exists")),
        Ok(false) => Ok(()),
        Err(error) => Err(PiError::other(format!("stat archive destination: {error}"))),
    }
}

/// Checks that `path` is exactly `<root>/<workspace key>/<name>.jsonl`.
///
/// Returns the cleaned root, the canonical workspace, the basename and the
/// cleaned candidate path.
fn validate_listed_candidate_path(
    root: &Path,
    workspace: &str,
    path: &Path,
) -> Result<(String, String, String, String), PiError> {
    let root_path = fsops::clean_go_path(&absolute(root)?);
    let workspace_canonical = fsops::canonical_workspace(Path::new(workspace))?;
    let key = fsops::workspace_key_of_canonical(&workspace_canonical);
    let candidate_path = fsops::clean_go_path(&absolute(path)?);
    let basename = candidate_path
        .rsplit('/')
        .next()
        .unwrap_or_default()
        .to_owned();
    let expected = format!("{root_path}/{key}/{basename}");
    if candidate_path != expected
        || basename == "."
        || basename == ".."
        || !basename.ends_with(".jsonl")
    {
        return Err(PiError::invalid(
            "listed session candidate is outside the expected workspace directory",
        ));
    }
    Ok((root_path, workspace_canonical, basename, candidate_path))
}

/// Go's `filepath.Abs`: the path joined to the working directory when relative.
fn absolute(path: &Path) -> Result<String, PiError> {
    if path.is_absolute() {
        return Ok(path.to_string_lossy().into_owned());
    }
    let working = std::env::current_dir()
        .map_err(|error| PiError::other(format!("resolve session root path: {error}")))?;
    Ok(working.join(path).to_string_lossy().into_owned())
}

fn open_listed_prepared_session_file(
    root: &str,
    workspace: &str,
    basename: &str,
) -> Result<(File, Metadata), PiError> {
    let root_dir = open_session_root_no_follow(Path::new(root))?;
    let Some((workspace_dir, _)) =
        open_workspace_session_directory_no_follow(&root_dir, Path::new(root), workspace)?
    else {
        return Err(PiError::invalid(
            "listed session directory no longer exists",
        ));
    };
    let file =
        fsops::open_at_no_follow(&workspace_dir, basename, libc::O_RDWR).map_err(|error| {
            if fsops::is_eloop(&error) {
                PiError::invalid("listed session file is a symlink")
            } else {
                PiError::other(format!("open listed session file: {error}"))
            }
        })?;
    let metadata = file
        .metadata()
        .map_err(|error| PiError::other(format!("stat listed session file: {error}")))?;
    if !metadata.is_file() {
        return Err(PiError::invalid(
            "listed session file is not a regular file",
        ));
    }
    Ok((file, metadata))
}

fn open_prepared_session_file_no_follow(path: &Path) -> Result<(File, Metadata), PiError> {
    let file = fsops::open_no_follow(path, libc::O_RDWR).map_err(|error| {
        if fsops::is_eloop(&error) {
            PiError::invalid("session file is a symlink")
        } else {
            PiError::other(format!("open session file: {error}"))
        }
    })?;
    let metadata = file
        .metadata()
        .map_err(|error| PiError::other(format!("stat session file: {error}")))?;
    if !metadata.is_file() {
        return Err(PiError::invalid("session file is not a regular file"));
    }
    Ok((file, metadata))
}

fn same_prepared_metadata(prepared: &SessionInfo, current: &SessionInfo) -> bool {
    prepared.id == current.id
        && prepared.cwd == current.cwd
        && prepared.profile == current.profile
        && prepared.provider == current.provider
        && prepared.model == current.model
        && prepared.created == current.created
}
